File: cpp/ThingThroughputProtobuf/src/ThroughputReader.cpp
/*                         ADLINK Edge SDK
 *
 *   This software and documentation are Copyright 2018 to 2020 ADLINK
 *   Technology Limited, its affiliated companies and licensors. All rights
 *   reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

/**
 * This is a simple throughput application measuring obtainable throughput using the thingSDK
 *
 */
#include <atomic>
#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <map>
#include <signal.h>
#include <sstream>
#include <stdlib.h>
#include <thread>
#include <time.h>

#include <IoTDataThing.hpp>
#include <JSonThingAPI.hpp>
#include <ThingAPIException.hpp>
#include "include/cxxopts.hpp"
#include "include/utilities.h"

#include "definitions/ThroughputTagGroup.dr.h"

using namespace std;
using namespace com::adlinktech::datariver;
using namespace com::adlinktech::iot;
using namespace com::adlinktech::example::protobuf;

unsigned long long sampleCount = 0;
unsigned long long startCount = 0;
static Timepoint currentTime = Timepoint();
static Timepoint startTime = Timepoint();
static unsigned long long bytesReceived = 0;
static unsigned long long outOfOrderCount = 0;
static unsigned long long batchCount = 0;
static unsigned long long batchMaxSize = 0;

static void showSummary() {
    // Output totals and averages
    if (batchCount > 0) {
        double deltaTime = (double)Duration<Microseconds>(currentTime - startTime) / US_IN_ONE_SEC;
        cout << endl << fixed
                    << "Total received: " << sampleCount << " samples, " << bytesReceived << " bytes" << endl
                    << "Out of order: " << outOfOrderCount << " samples" << endl
                    << "Average transfer rate: "
                            << setprecision(0) << (double)sampleCount / deltaTime << " samples/s, "
                            << setprecision(2) << ((double)bytesReceived / BYTES_PER_SEC_TO_MEGABITS_PER_SEC) / deltaTime << " Mbit/s"
                    << endl
                    << "Average sample-count per batch: " << sampleCount / batchCount << ", maximum batch-size: " << batchMaxSize << endl;
    }
}

#ifdef _WIN32
static bool ctrlHandler(DWORD fdwCtrlType)
{
    showSummary();
    ExitProcess(1);
}
#else
static void ctrlHandler(int fdwCtrlType)
{
    showSummary();
    exit(1);
}
#endif

class ThroughputReader {
private:

    string m_thingPropertiesUri;
    DataRiver m_dataRiver = createDataRiver();
    ThingEx m_thing = createThing();

    DataRiver createDataRiver() {
        return DataRiver::getInstance();
    }

    ThingEx createThing() {
        // Create and Populate the TagGroup registry with JSON resource files.
        ThroughputHelper::registerWithDataRiver(m_dataRiver);

        // Create and Populate the ThingClass registry with JSON resource files.
        JSonThingClassRegistry tcr;
        tcr.registerThingClassesFromURI("file://definitions/ThingClass/com.adlinktech.example.protobuf/ThroughputReaderThingClass.json");
        m_dataRiver.addThingClassRegistry(tcr);

        // Create a Thing based on properties specified in a JSON resource file.
        JSonThingProperties tp;
        tp.readPropertiesFromURI(m_thingPropertiesUri);

        return m_dataRiver.createThing(tp);
    }

public:
    ThroughputReader(string thingPropertiesUri) :
            m_thingPropertiesUri(thingPropertiesUri) {
        cout << "Throughput reader started" << endl;
    }

    ~ThroughputReader() {
        m_dataRiver.close();
        cout << "Throughput reader stopped" << endl;
    }

    int run(unsigned long pollingDelay, unsigned long runningTime) {
        unsigned long long prevCount = 0;
        unsigned long long prevReceived = 0;
        unsigned long long samplesInBatch = 0;
//        unsigned long long payloadSize = 0;
        unsigned long long deltaReceived;
        unsigned long long deltaTime;
        Timepoint prevTime = Timepoint();
        IOT_UINT64 receivedSequenceNumber;
        IOT_UINT64 lastReceivedSequenceNumber;
        unsigned long long payloadSize = 0;
        bool firstSample = true;

        cout << "Waiting for samples..." << endl;

        // Loop through until the runningTime has been reached (0 = infinite)
        // each cycle is 1 second
        unsigned long long cycles = 0;

        while (!stop && (runningTime == 0 || cycles < runningTime))
        {
            if (pollingDelay > 0) {
                this_thread::sleep_for(chrono::milliseconds(pollingDelay));
            }

            // New batch
            batchCount++;
            samplesInBatch = sampleCount;

            // Take samples and iterate through them
            VLoanedDataSamples samples = m_thing.read("ThroughputInput", BLOCKING_TIME_INFINITE);
            // in loops, take care to declare auto variables 'auto&', otherwise, you will
            // get a copy of the data
            for(auto& sample : samples) {
                if (sample.getFlowState() == FlowState::ALIVE) {
                    Throughput data;
                    sample.get(data);
                    receivedSequenceNumber = data.sequencenumber();
                    payloadSize = data.sequencedata().size();
                    bytesReceived += payloadSize;
                    sampleCount++;

                    if (firstSample) {
                        lastReceivedSequenceNumber = receivedSequenceNumber - 1;
                        firstSample = false;
                    }

                    // Check that the sample is the next one expected
                    if (receivedSequenceNumber != lastReceivedSequenceNumber + 1) {
                        outOfOrderCount += (receivedSequenceNumber - (lastReceivedSequenceNumber + 1 ));
                    }

                    // Keep track of last received seq nr
                    lastReceivedSequenceNumber = receivedSequenceNumber;
                } else {
                    cout << "Writer flow purged, stop reader" << endl;
                    stop = true;
                }
            }

            if (!stop) {
                currentTime = Clock::now();
                if (Duration<Microseconds>(currentTime - prevTime) > US_IN_ONE_SEC) {
                    // If not the first iteration
                    if (prevTime.time_since_epoch() != Timepoint().time_since_epoch()) {
                        // Calculate the samples and bytes received and the time passed since the  last iteration and output
                        deltaReceived = bytesReceived - prevReceived;
                        deltaTime = Duration<Microseconds>(currentTime - prevTime) / US_IN_ONE_SEC;

                        cout << fixed
                                    << "Payload size: " << payloadSize << " | "
                                    << "Total: " << setw(9) << right << sampleCount << " samples, "
                                    << setw(12) << right << bytesReceived << " bytes | "
                                    << setw(6) << right << "Out of order: " << outOfOrderCount << " samples | "
                                    << "Transfer rate: " << setw(7) << right << setprecision(0) << (double)(sampleCount - prevCount) / deltaTime << " samples/s, "
                                        << setw(9) << right << setprecision(2) << ((double)deltaReceived / BYTES_PER_SEC_TO_MEGABITS_PER_SEC) / deltaTime << " Mbit/s"
                                    << endl;

                        cycles++;
                    }
                    else
                    {
                        // Set the start time if it is the first iteration
                        startTime = currentTime;
                    }

                    // Update the previous values for next iteration
                    prevReceived = bytesReceived;
                    prevCount = sampleCount;
                    prevTime = currentTime;
                }

                // Update max samples per batch
                samplesInBatch = sampleCount - samplesInBatch;
                if (samplesInBatch > batchMaxSize) {
                    batchMaxSize = samplesInBatch;
                }
            }
        }

        showSummary();

        return 0;
    }
};


static void GetCommandLineParameters(int argc, char *argv[],
        unsigned long& pollingDelay,
        unsigned long& runningTime
) {
    cxxopts::Options options("ThroughputReader", "ADLINK ThingSDK ThroughputReader");
    try {
        options.add_options()
            ("p,polling-delay", "Polling delay (milliseconds)", cxxopts::value<unsigned long>()->default_value("0"))
            ("r,running-time", "Running time (seconds, 0 = infinite)", cxxopts::value<unsigned long>()->default_value("0"))
            ("h,help", "Print help")
            ;
        options.parse_positional({"polling-delay", "running-time", "other"});
        options.positional_help("[polling-delay] [running-time]");

        auto cmdLineOptions = options.parse(argc, argv);

        if (cmdLineOptions.count("help")) {
            cout << options.help({""}) << endl;
            exit(0);
        }

        pollingDelay = cmdLineOptions["p"].as<unsigned long>();
        runningTime = cmdLineOptions["r"].as<unsigned long>();
    }
    catch (cxxopts::OptionException e) {
        cerr << e.what() << endl << endl;
        cout << options.help({""}) << endl;
        exit(1);
    }catch (const std::domain_error e1) {
        cerr << e1.what() << endl << endl;
        cout << options.help({""}) << endl;
        exit(1);
    }
}


int main(int argc, char *argv[]) {
    // Get command line parameters
    unsigned long pollingDelay;
    unsigned long runningTime;
    GetCommandLineParameters(argc, argv, pollingDelay, runningTime);

    // Register handler for Ctrl-C
    registerControlHandler();

    try
    {
        return ThroughputReader("file://./config/ThroughputReaderProperties.json")
            .run(pollingDelay, runningTime);
    }
    catch (ThingAPIException e)
    {
        cerr << "An unexpected error occurred: " << e.what() << endl;
    }catch(std::exception& e1){
        cerr << "An unexpected error occurred: " << e1.what() << endl;
    }

    unregisterControlHandler();
}

File: python/ThingThroughput/throughput_reader.py
#!/usr/bin/env python

#
#                           ADLINK Edge SDK
#  
#     This software and documentation are Copyright 2018 to 2020 ADLINK
#     Technology Limited, its affiliated companies and licensors. All rights
#     reserved.
#  
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#  
#         http://www.apache.org/licenses/LICENSE-2.0
#  
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#

'''
This is a simple throughput application measuring obtainable throughput using the thingSDK

Updated the default running-time to 60 seconds as KeyboardInterrupt does not work with the OpenSplice on Windows.
'''

from __future__ import print_function
import argparse
import sys
import six
import time
import os
from enum import Enum
from adlinktech.datariver import DataRiver, JSonTagGroupRegistry, JSonThingClassRegistry, JSonThingProperties, FlowState
from adlinktech.datariver import IotValue, IotNvp, IotNvpSeq, BLOCKING_TIME_INFINITE

US_IN_ONE_SEC = 1000000
BYTES_PER_SEC_TO_MEGABITS_PER_SEC = 125000

# Convert python returned current time expressed in seconds(as floating point number) to microseconds
# Note: On Windows, time.time() does not provide time with a better precision than 1 second
if six.PY2:
    current_time_microseconds = lambda: int(round(time.time() * 1000000))
else:
    current_time_microseconds = lambda: int(round(time.perf_counter() * 1000000))

sample_count = 0
current_time = current_time_microseconds()
start_time = current_time_microseconds()
bytes_received = 0
out_of_order_count = 0
batch_count = 0
batch_max_size = 0

stop = False

def show_summary():
    # Output totals and averages
    global sample_count, current_time, start_time, bytes_received, out_of_order_count, batch_count, batch_max_size
    if batch_count > 0:
        delta_time = (current_time - start_time) / US_IN_ONE_SEC
        if delta_time > 0:
            print(os.linesep + 'Total received: {} samples, {} bytes'.format(sample_count, bytes_received) + os.linesep +
                  'Out of order: {} samples'.format(out_of_order_count) + os.linesep +
                  'Average transfer rate: {:.0f} samples/s, {:.2f} Mbit/s'
                  .format(float(sample_count) / delta_time, (float(bytes_received) / BYTES_PER_SEC_TO_MEGABITS_PER_SEC) / delta_time)
                   + os.linesep +
                   'Average sample-count per batch: {}, maximum batch-size: {}'.format(int(sample_count / batch_count), batch_max_size))
        else:
            print(os.linesep + 'Total received: {} samples, {} bytes'.format(sample_count, bytes_received) + os.linesep +
                  'Out of order: {} samples'.format(out_of_order_count) + os.linesep +
                  'Average transfer rate: {:.0f} samples/s, {:.2f} Mbit/s'
                  .format(0.0, 0.0)
                   + os.linesep +
                   'Average sample-count per batch: {}, maximum batch-size: {}'.format(int(sample_count / batch_count), batch_max_size))


class ThroughputReader(object):
     
    # Initializing
    def __init__(self, thing_properties_uri):
        self._thing_properties_uri = thing_properties_uri
        
        self._datariver = None
        self._thing = None

    # Enter the runtime context related to the object
    def __enter__(self):
        self._datariver = DataRiver.get_instance()
        self._thing = self.create_thing()
        
        print('Throughput reader started')
        
        return self

    # Exit the runtime context related to the object
    def __exit__(self, exc_type, exc_value, exc_traceback):
        if self._datariver is not None:
            self._datariver.close()
        print('Throughput reader stopped')

    def create_thing(self):
        # Create and Populate the TagGroup registry with JSON resource files.
        tgr = JSonTagGroupRegistry()
        tgr.register_tag_groups_from_uri('file://definitions/TagGroup/com.adlinktech.example/ThroughputTagGroup.json')
        self._datariver.add_tag_group_registry(tgr) 
  
        # Create and Populate the ThingClass registry with JSON resource files.
        tcr = JSonThingClassRegistry()
        tcr.register_thing_classes_from_uri('file://definitions/ThingClass/com.adlinktech.example/ThroughputReaderThingClass.json')
        self._datariver.add_thing_class_registry(tcr)
  
        # Create a Thing based on properties specified in a JSON resource file.
        tp = JSonThingProperties()
        tp.read_properties_from_uri(self._thing_properties_uri)
        
        return self._datariver.create_thing(tp)
    
    
    
    def run(self, polling_delay, running_time):
        prev_count = 0
        prev_received = 0
        samples_in_batch = 0
        payload_size = 0
        delta_received = None
        delta_time = None
        first_time = True
        prev_time = current_time_microseconds()
        received_sequence_number = None
        last_received_sequence_number = None
        received_data = None
        seq_nr_found = False
        first_sample = True
        
        print('Waiting for samples...')
        
        # Loop through until the runningTime has been reached (0 = infinite)
        # each cycle is 1 second
        cycles = 0
        
        global stop, sample_count, current_time, start_time, bytes_received, out_of_order_count, batch_count, batch_max_size
        
        while not stop and (running_time == 0 or cycles < running_time):
            if polling_delay > 0:
                time.sleep(float(polling_delay)/1000.0)
            
            # Take samples and iterate through them
            samples = self._thing.read_iot_nvp('ThroughputInput', BLOCKING_TIME_INFINITE)
            
            # New batch
            batch_count += 1
            samples_in_batch = sample_count
            
            for sample in samples:
                if sample.flow_state == FlowState.ALIVE:
                    data = sample.data
                    
                    # find the message, stored in the name-value-pair with name 'name':
                    seq_nr_found = False
                    for nvp in data:
                        if nvp.name == 'sequencenumber':
                            received_sequence_number = nvp.value.uint64
                            seq_nr_found = True
                        elif nvp.name == 'sequencedata':
                            # Add the sample payload size to the total received
                            received_data = nvp.value.byte_seq
                            payload_size = received_data.size()
                            bytes_received += payload_size + 8 # add 8 bytes for sequence number field
                    
                    if seq_nr_found:
                        # Increase sample count
                        sample_count += 1
                        
                        if first_sample:
                            last_received_sequence_number = received_sequence_number - 1
                            first_sample = False
                        
                        # Check that the sample is the next one expected
                        if received_sequence_number != last_received_sequence_number + 1:
                            out_of_order_count += (received_sequence_number - (last_received_sequence_number + 1))
                        
                        # Keep track of last received seq nr
                        last_received_sequence_number = received_sequence_number
                else:
                    print('Writer flow purged, stop reader')
                    stop = True
            
            if not stop:
                current_time = current_time_microseconds()
                if (current_time - prev_time) > US_IN_ONE_SEC:
                    # If not the first iteration
                    if not first_time:
                        # Calculate the samples and bytes received and the time passed since the  last iteration and output
                        delta_received = bytes_received - prev_received
                        delta_time = (current_time - prev_time) / US_IN_ONE_SEC
                        
                        print('Payload size: {} | '.format(payload_size) +
                              'Total: {:>9} samples, '.format(sample_count) +
                              '{:>12} bytes | '.format(bytes_received) +
                              'Out of order: {:>6} samples | '.format(out_of_order_count) +
                              'Transfer rate: {:>7.0f} samples/s, '.format(float(sample_count - prev_count) / delta_time) +
                              '{:>9.2f} Mbit/s'.format((float(delta_received) / BYTES_PER_SEC_TO_MEGABITS_PER_SEC) / delta_time))
                        
                        cycles += 1
                    else:
                        # Set the start time if it is the first iteration
                        first_time = False
                        start_time = current_time
                    
                    # Update the previous values for next iteration
                    prev_received = bytes_received
                    prev_count = sample_count
                    prev_time = current_time
                
                # Update max samples per batch
                samples_in_batch = sample_count - samples_in_batch
                if samples_in_batch > batch_max_size:
                    batch_max_size = samples_in_batch
                
        show_summary()
        
        return 0
    
def get_command_line_parameters():
    parser = argparse.ArgumentParser(description='ADLINK ThingSDK ThroughputReader')
    parser.add_argument('-p', '--polling-delay', type=int,
                        help='Polling delay (milliseconds)', default=0)
    parser.add_argument('-r', '--running-time', type=int,
                        help='Running time (seconds, 0 = infinite)',
                        default=0)
    
    args = parser.parse_args()
    
    return args.polling_delay, args.running_time

def main():
    # Get command line parameters
    polling_delay, running_time = get_command_line_parameters()
    
    try:
        throughput_rd = ThroughputReader('file://./config/ThroughputReaderProperties.json')
        # The 'with' statement supports a runtime context which is implemented
        # through a pair of methods executed (1) before the statement body is
        # entered (__enter__()) and (2) after the statement body is exited (__exit__())
        with throughput_rd as thp_rd:
            thp_rd.run(polling_delay, running_time)
    except Exception as e:
        print('An unexpected error occurred: {}'.format(e))
    except KeyboardInterrupt:
        show_summary()
        sys.exit(1)

if __name__ == '__main__':
    main()


